File: bitmap/bitmap.c
#include <stdio.h>   
#include <stdlib.h>   
#include <string.h>   
#include "bitmap.h"   
  
int bitmap_init(bitmap* b, int size, int start){  
    b->p = (unsigned char *)malloc((size/8+1)*sizeof(char));  
    if(b->p == NULL)  return 0;  
    b->base = start;  
    b->size = size/8+1;  
    memset(b->p , 0x0, b->size);  
    return 1;  
}
  
int bitmap_set(bitmap* b,int index)  
{
    int quo = (index-b->base) >> 3 ;  
    int remainder = (index-b->base)&7;  
    unsigned char x = (0x1<<remainder);  
    if( quo > b->size)  
        return 0;  
    b->p[quo] |= x;  
    printf("%d %d %d\n",index,quo,remainder);
    return 1;   
}  
  
inline int bitmap_get(bitmap* b,int i)  
{  
    int quo = (i)>>3 ;  
    int remainder = (i)&7;  
    unsigned char x = (0x1<<remainder);  
    unsigned char res;  
    if( quo > b->size)  
        return -1;  
    res = b->p[quo] & x;  
    return res > 0 ? 1 : 0;   
}  

//intersection 
bitmap* bitmap_inter(bitmap* b1,bitmap* b2)  
{  
    bitmap* b = (bitmap*)malloc(sizeof(bitmap));
    bitmap_init(b,BITMAP_SIZE, 0);
    int i;
    for(i=0;i<b->size;i++){
        b->p[i] = (b1->p[i] & b2->p[i]);
    }
    return b;
}

int bitmap_data(bitmap* b,int index)  
{  
    return (index + b->base);  
}  
  
int bitmap_free(bitmap* b)  
{  
    free(b->p);  
    free(b);  
    return 0;
}

int bitmap_count1(bitmap* b){
    int total = 0;
    int a = 0;
    int i;
    for(i=0;i<b->size;i++){
        a = (int)(b->p[i]);
        while(a){
            total += 1;
            a = a & (a-1);
        }
    }
    return total;
}

File: bitmap/bitmap.go
package main

import (
    "log"
    "fmt"
    "time"
)

const(
    BITMAPSIZE = 1024*1024*200
)

type Bitmap struct{
    Base    uint8  `sizeof the underlay type,say 8`
    BaseExp uint8  `log2(Base),exponent of the Base,say 3`
    Size    int64  `len(Bits),num of the underlay type,say BITMAPSIZE/Base`
    Bits    []int64 `bitmap holds the values,underlay-type:byte int64`
}

func NewBitmap()*Bitmap{
    var b = Bitmap{}
    b.Base = 64
    b.BaseExp = 8
    b.Size = BITMAPSIZE/int64(b.Base)
    b.Bits = make([]int64,b.Size)
    return &b
}

func(b *Bitmap)setBit(index int64)int{
    quo := index >> b.BaseExp
    remainder := uint8(index & int64(b.Base-1))
    if quo > b.Size{
        return -1
    }
    b.Bits[quo] |= (0x1<<remainder)
    return 0
}

func(b *Bitmap)getBit(index int64)int{
    quo := index >> b.BaseExp
    remainder := uint8(index & int64(b.Base-1))
    if quo > b.Size{
        return -1
    }

    res := b.Bits[quo] & (0x1<<remainder)
    if res >0{
        return 1
    }else{
        return 0
    }
}

func(b *Bitmap)inter(c *Bitmap)*Bitmap{
    var i int64
    for i =0;i<b.Size;i++{
        b.Bits[i] &= c.Bits[i]
    }
    return b
}

func(b *Bitmap)union(c *Bitmap)*Bitmap{
    var i int64
    for i=0;i<b.Size;i++{
        b.Bits[i] |= c.Bits[i]
    }
    return b
}

func(b *Bitmap)count1()int64{
    var (
        i int64
        p int64
        total int64
    )

    for i=0;i<b.Size;i++{
        p = int64(b.Bits[i])
        for p >0{
            total += 1
            p = p & (p-1)
        }
    }
    return total
}

func main(){
    log.SetFlags(23)
    b := NewBitmap()
    c := NewBitmap()

    for _,v:= range []int64{5,8,7,6,3,1,10,19}{
        b.setBit(v)
    }

    for _,v:= range []int64{5,8,7,6,3,1}{
        c.setBit(v)
    }

    a := NewBitmap()
    t:= time.Now()
    a.union(b).inter(c)
    log.Println(time.Since(t))
    for i:=0;i<20;i++{
        fmt.Print(b.getBit(int64(i)))
    }
    fmt.Println()
    for i:=0;i<20;i++{
        fmt.Print(c.getBit(int64(i)))
    }
    fmt.Println()
    for i:=0;i<20;i++{
        fmt.Print(a.getBit(int64(i)))
    }
    fmt.Println()
    fmt.Println(a.count1())
    fmt.Println()
}
